File: 1980-find-unique-binary-string/1980-find-unique-binary-string.py
class Solution:
    def findDifferentBinaryString(self, nums: List[str]) -> str:
        def dfs(curr):
            if len(curr) == n:
                if curr not in nums:
                    return curr
                return ""
            
            add_zero = dfs(curr + "0")
            if add_zero:
                return add_zero

            add_one = dfs(curr + "1")
            return add_one

        n = len(nums)
        nums = set(nums)
        return dfs("")

# array, backtracking, hash map, string
# time: O(2^n)
# space: O(n)

File: 1980-find-unique-binary-string/1980-find-unique-binary-string.go
func findDifferentBinaryString(nums []string) string {
    n := len(nums)
    m := map[string]bool{}
    for _, num := range nums {
        m[num] = true
    }

    var dfs func(s string) string
    dfs = func(s string) string {
        if len(s) == n {
            if !m[s] {
                return s
            }
            return ""
        }

        addZero := dfs(s + "0")
        if addZero != "" {
            return addZero
        }

        addOne := dfs(s + "1")
        return addOne
    }

    return dfs("")
}

// array, backtracking, hash map, string
// time: O(2^n)
// space: O(n)